lib/model/autocomplete.dart

```diff
@@ -244,11 +246,108 @@ class MentionAutocompleteView extends ChangeNotifier {
+    final countB = store.recentDmConversationsView.getRecipientCount(userB.userId);
+
+    if (countA > countB) {
+      return -1;
+    } else if (countB > countA) {
+      return 1;
+    }
+
+    if (!userA.isBot && userB.isBot) {
+      return -1;
+    } else if (userA.isBot && !userB.isBot) {
+      return 1;
+    }
+
+    if (userA.fullName.toLowerCase().compareTo(
+      userB.fullName.toLowerCase()) < 0) {
+      return -1;
+    }
+    return 1;
+  }
+
+  int compareByRelevance({
+    required User userA,
+    required User userB,
+    required int Function(User, User) tertiaryCompare,
+  }) {
+    // TODO: give preference to "all", "everyone" or "stream".
+
+    // TODO: give preference to subscribed users first
+
+    // give preference to direct message partners if both (are)/(are not) subscribers
+    final aIsPartner = store.recentDmConversationsView.isPartner(userA.userId);
+    final bIsPartner = store.recentDmConversationsView.isPartner(userB.userId);
+
+    if (aIsPartner && !bIsPartner) {
+      return -1;
+    } else if (bIsPartner && !aIsPartner) {
+      return 1;
+    }
+
+    return tertiaryCompare(userA, userB);
+  }
+
+  List<User> sortByRelevance({
+    required List<User> users,
+    required int? currentStreamId,
+    required String? currentTopic,
+  }) {
+    // Sorting for DMs narrow
+    if (currentStreamId == null) {
+      users.sort((userA, userB) => compareByRelevance(
+        userA: userA,
+        userB: userB,
+        tertiaryCompare: compareByDms));
+    } else { // Sorting for stream/topic narrow
+      users.sort((userA, userB) => compareByRelevance(
+        userA: userA,
+        userB: userB,
+        tertiaryCompare: (userA, userB) => store.recentSenders.compareByRecency(
+          userA: userA,
+          userB: userB,
+          streamId: currentStreamId,
+          topic: currentTopic)));
+    }
+    return users;
+  }
+
+  void _sortUsers() {
+    final users = store.users.values.toList();
+    switch(narrow) {
+      case StreamNarrow():
+        _sortedUsers = sortByRelevance(
+            users: users,
+            currentStreamId: (narrow as StreamNarrow).streamId,
+            currentTopic: null);
+      case TopicNarrow():
+        _sortedUsers = sortByRelevance(
+            users: users,
+            currentStreamId: (narrow as TopicNarrow).streamId,
+            currentTopic: (narrow as TopicNarrow).topic);
+      case DmNarrow():
+        _sortedUsers = sortByRelevance(
+            users: users,
+            currentStreamId: null,
+            currentTopic: null);
+      case AllMessagesNarrow():
+        _sortedUsers = []; // TODO: sort (maybe in future)
+    }
+  }
+
   Future<List<MentionAutocompleteResult>?> _computeResults(MentionAutocompleteQuery query) async {
     final List<MentionAutocompleteResult> results = [];
-    final Iterable<User> users = store.users.values;
 
-    final iterator = users.iterator;
+    if (_sortedUsers == null) {
+      _sortUsers();
+    }
```

Comment:
Two options here:
1. Sort all users one time at the start of the query, then filter them??
2. Filter all users, then sort them each time according to the query??

In both cases, filtering is done on the same full-length list,
but sorting is done in case:
  1) one time on a full-length list
  2) each time on a smaller-length filtered list

The web is using the second approach! Which one should we choose?